pyquil/_memory.py

```diff
@@ -0,0 +1,56 @@
+from dataclasses import dataclass, field
+from typing import Dict, Sequence, Union
+
+from rpcq.messages import ParameterAref
+
+
+@dataclass
+class Memory:
```

Comment:
pyQuil uses this Memory class only for writing program input. `QAMExecutionResult#memory` on the other hand is exclusively for reading memory.

Symmetry might not be worthwhile, but I would at least semantically suggest the function more specifically in the class or attribute names - eg `ReadMemory`, `Program#read_memory` or something of that ilk.

Reply:
nailed it - I had actually addressed this in a review comment I never submitted.

That's why this is private and abstracted away in the interfaces of the classes in which it's used. Once this evolves to be RW memory, we can make it public and expose it directly.

Reply:
To follow up, there's at least one reason why this would be difficult to do now. Exposing `ro` and other readout data as "memory" is itself a lie to the user. When the user sees an object of shape `(1000, 1)` returned from a single-qubit readout with 1000 shots, that's not really memory in the true sense; that's classified readout data masquerading as a memory region. However, it's how our backend behaves right now so we're stuck with it in the near term.

There's another trick, too - programs have memory which is defined by the size and types of their declared regions. The 'result' memory though is not only defined by the number of shots (as well as size and type of memory) but its size is defined on the service side rather than the data handled here by `Memory`. This is a tractable problem, but it's something I feel we'd rather handle together with the previous problem in one well-documented change.

Thoughts?

Reply:
I think the current state of things, with `readout_data` defined on `QAMExecutionResult` and a private `Memory` class is the best approach for now.

I don't entirely see a path forward for accessing QPU program memory through a pointer (ie as anything but readout data), but maybe I'm misunderstanding your point.